README.md

# Construct
[![Build Status](https://github.com/jonathantorres/construct/actions/workflows/tests.yml/badge.svg?branch=master)](https://github.com/jonathantorres/construct/actions/workflows/tests.yml)
[![Build Status](https://ci.appveyor.com/api/projects/status/github/jonathantorres/construct?svg=true)](https://ci.appveyor.com/project/jonathantorres/construct)
[![Version](http://img.shields.io/packagist/v/jonathantorres/construct.svg?style=flat)](https://packagist.org/packages/jonathantorres/construct)
[![PDS Skeleton](https://img.shields.io/badge/pds-skeleton-blue.svg?style=flat)](https://github.com/php-pds/skeleton)

A PHP project/micro-package generator for [PDS](http://php-pds.com/) compliant projects or micro-packages.

## Installation
Construct should be installed globally through composer.

```bash
composer global require jonathantorres/construct
```

Make sure that `~/.composer/vendor/bin` is on your `$PATH`. This way the `construct` executable can be located.

## Assumptions

As Construct utilizes Composer's CLI, it's assumed that Composer is installed. When using the option to initialize an empty Git repo (i.e. `--git` or `-g`) it's also assumed that Git is installed.

## Usage
Just run `construct generate` with your `vendor/package` declaration and it will construct a basic PHP project into the `package` directory. For example, if you run `construct generate jonathantorres/logger` it will generate a basic project structure inside the `logger` folder.

```bash
construct generate jonathantorres/logger
```

The generated project structure will look like the following `tree` excerpt. Files and directories in parentheses are optional.
```
├── logger
│   ├── CHANGELOG.md
│   ├── (CONDUCT.md)
│   ├── composer.json
│   ├── composer.lock
│   ├── CONTRIBUTING.md
│   ├── (.appveyor.yml)
│   ├── (.editorconfig)
│   ├── (.env)
│   ├── (.env.example)
│   ├── (.git)
│   │   └── ...
│   ├── .gitattributes
│   ├── (.github)
│   │   ├── CONTRIBUTING.md
│   │   ├── ISSUE_TEMPLATE.md
│   │   └── PULL_REQUEST_TEMPLATE.md
│   ├── .gitignore
│   ├── .gitmessage
│   ├── (.lgtm)
│   ├── (bin)
│   │   └── cli-script
│   ├── LICENSE.md
│   ├── (MAINTAINERS)
│   ├── (.php_cs)
│   ├── (phpunit.xml.dist)
│   ├── README.md
│   ├── (docs)
│   │   └── index.md
│   ├── src
│   │   └── Logger.php
│   ├── tests
│   │   └── LoggerTest.php
│   ├── .travis.yml
│   ├── (Vagrantfile)
│   └── vendor
│           └── ...
```

This is a good starting point. You can continue your work from there.

## Select testing framework
The `--test-framework` or `--test` option will allow you to select a testing framework. One of the following is available at the moment: `phpunit`, `phpspec`, `codeception` or `behat`. `phpunit` is currently the default.

```bash
construct generate jonathantorres/logger --test-framework=codeception
```

You can also use the short option `-t`.

```bash
construct generate jonathantorres/logger -t codeception
```

## Select license
The `--license` option will allow you to select a license for the project to construct. One of the following is
available at the moment: `MIT`, `Apache-2.0`, `GPL-2.0` or `GPL-3.0`. `MIT` is currently the default.

```bash
construct generate jonathantorres/logger --license=Apache-2.0
```

You can also use the short option `-l`.

```bash
construct generate jonathantorres/logger -l Apache-2.0
```

## Specify a namespace
The `--namespace` option will allow you to specify a namespace for the project to construct. This is totally optional. By default construct will use the package name as the namespace.

```bash
construct generate jonathantorres/logger --namespace=JonathanTorres\\Projects\\Logger
```

You can also use the short option `-s`.
```bash
construct generate jonathantorres/logger -s JonathanTorres\\Projects\\Logger
```

## Specify a CLI framework
The optional `--cli-framework` option will allow you to specify a CLI framework for the project to construct, while also creating a bin directory with an initial CLI script in it, adding a bin key in the project's composer.json, and an initial [AppVeyor](https://www.appveyor.com) configuration. When the option has been set without a CLI composer package the `symfony/console` package will be used per default. There's no short option available.

```bash
construct generate jonathantorres/logger --cli-framework=zendframework/zend-console
```

## Specify PHP version
The `--php` option will allow you to specify the minimum required PHP version that your project will support. Construct will use the currently installed version if not specified.

```bash
construct generate jonathantorres/logger --php=5.6.31
```

## Specify Composer keywords
The optional `--keywords` option will allow you to specify a comma separated list of [Composer keywords](https://getcomposer.org/doc/04-schema.md#keywords).

```bash
construct generate jonathantorres/logger --keywords=log,logging
```

You can also use the short option `-k`.

```bash
construct generate jonathantorres/logger -k=log,logging
```

## Initialize an empty Git repo?
The `--git` option will allow you to initialize an empty Git repository inside the constructed project.

```bash
construct generate jonathantorres/logger --git
```

You can also use the short option `-g`.

```bash
construct generate jonathantorres/logger -g
```

## Generate a PHP Coding Standards Fixer configuration?
The `--phpcs` option will generate a [PHP Coding Standards Fixer](http://cs.sensiolabs.org/) configuration within the constructed project and add a Travis CI script for validation during builds. The generated `.php_cs` configuration defaults to the `PSR-2` coding style guide.

```bash
construct generate jonathantorres/logger --phpcs
```

You can also use the short option `-p`.

```bash
construct generate jonathantorres/logger -p
```

## Generate a Vagrantfile?
The `--vagrant` option will generate a basic [Vagrantfile](https://docs.vagrantup.com/v2/vagrantfile/index.html) within the constructed project, defaulting to the output of a `vagrant init` call plus a minimal [vagrant-cachier](http://fgrehm.viewdocs.io/vagrant-cachier) plugin configuration. There's no short option available.

```bash
construct generate jonathantorres/logger --vagrant
```

## Generate an EditorConfig configuration?
The `--editor-config` option will generate an [EditorConfig](http://editorconfig.org) configuration
within the constructed project.

```bash
construct generate jonathantorres/logger --editor-config
```

You can also use the short option `-e`.

```bash
construct generate jonathantorres/logger -e
```

## Generate .env enviroment files?
The `--env` option will generate [.env](https://github.com/vlucas/phpdotenv) environment files within the constructed project for keeping `sensitive` information out of it. There's no short option available.

```bash
construct generate jonathantorres/logger --env
```

## Generate LGTM configuration files?
The `--lgtm` option will generate [LGTM](https://lgtm.co) configuration files within the constructed project. There's no short option available.

```bash
construct generate jonathantorres/logger --lgtm
```

## Generate common GitHub files?
The `--github` option will generate common GitHub (i.e. `templates` and `docs`) files. There's no short option available.

```bash
construct generate jonathantorres/logger --github
```

## Generate GitHub template files?
The `--github-templates` option will generate [GitHub template](https://github.com/blog/2111-issue-and-pull-request-templates) files within the constructed project into a folder conventionally named `.github`. It also will move `CONTRIBUTING.md` into it. There's no short option available, but the option is implicated with the `github` option.

```bash
construct generate jonathantorres/logger --github-templates
```

## Generate GitHub documentation files?
The `--github-docs` option will generate [GitHub documentation](https://github.com/blog/2233-publish-your-project-documentation-with-github-pages) files within the constructed project into a folder conventionally named `docs`. There's no short option available, but the option is implicated with the `github` option.

```bash
construct generate jonathantorres/logger --github-docs
```

## Generate a Code of Conduct?
The `--code-of-conduct` option will generate a Code of Conduct file named `CONDUCT.md` within the constructed project and also add a reference to it in the generated `README.md`. The used Code of Conduct is adapted from the [Contributor Covenant](http://contributor-covenant.org), version 1.4. There's no short option available.

```bash
construct generate jonathantorres/logger --code-of-conduct
```

## Use a configuration for recurring settings
The `--config` option allows the usage of a configuration file in the `YAML` format. There are two ways to provide such a configuration file: One is to provide a specific file as an option argument, the other is to put a `.construct` configuration file in the home directory of your system. For the structure of a configuration file have a look at the [.construct](example/.construct) example file. When no configuration keys are provided for settings having a default value (i.e. `test-framework`, `license`, `php`) their default value is used.

```bash
construct generate jonathantorres/logger --config /path/to/config.yml
```

You can also use the short option `-c`.

```bash
construct generate jonathantorres/logger -c /path/to/config.yml
```

When there's a `.construct` configuration file in your home directory it will be used per default. If required it's possible to disable the usage via the `--ignore-default-config` option or the equivalent short option `-i`.

## Interactive Mode
This mode will ask you all the required (and optional) information to generate your project.

```bash
construct generate:interactive
```

## Running tests
``` bash
composer construct:test
```

## License
This library is licensed under the MIT license. Please see [LICENSE](LICENSE.md) for more details.

## Changelog
Please see [CHANGELOG](CHANGELOG.md) for more details.

## Contributing
Please see [CONTRIBUTING](CONTRIBUTING.md) for more details.


CHANGELOG.md
# Changelog

#### v3.0.0 `2023-01-21`
- `Added`
    - Support for PHP 8.0, 8.1 and 8.2 only.
    - Bump PHPUnit to `8.0`.

#### v2.1.0 `2022-08-28`
- `Added`
    - Support for PHP 7.3 and 7.4.

#### v2.0.0 `2018-04-01`
- `Breaking Changes`
    - Support for PHP7 only.
    - Bump Symfony dependencies to `3.0`.
    - Bump PHPUnit to `6.0` and Mockery to `1.0`.
- `Added`
    - Added Composer script descriptions to improve Composer script documentation when runtime Composer is >= `1.6.0`. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#215](https://github.com/jonathantorres/construct/issues/215).
    - Added Composer script descriptions for internal Composer scripts, bumps the by Construct required Composer version to `^1.6.0`. Done by [@raphaelstolt](https://github.com/raphaelstolt).

#### v1.16.0 `2017-10-26`
- `Added`
    - Added PHP version 7.2 to the available versions. Done by [@raphaelstolt](https://github.com/raphaelstolt).
- `Fixed`
    - Removed EOL PHP versions (i.e. `5.4` and `5.5`). Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#211](https://github.com/jonathantorres/construct/issues/211).
    - Added a guard to check if the project directory to be already exists. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#212](https://github.com/jonathantorres/construct/issues/212).
    - Generated AppVeyor configuration uses minor versions instead of ZIP archives. Done by [@raphaelstolt](https://github.com/raphaelstolt).
    - Minor improvements on GitHub template files. Done by [@piotr-zuralski](https://github.com/piotr-zuralski).

#### v1.15.2 `2017-09-18`
- `Fixed`
    - Major codebase refactor.

#### v1.15.1 `2017-08-28`
- `Fixed`
    - Use the default CLI framework `symfony/console` if the one specified has an invalid name. Done by [@jonathantorres](https://github.com/jonathantorres). See [1b297ca](https://github.com/jonathantorres/construct/commit/1b297ca8a9bd791111cb0aa8a5aeb87b4c9dbc4d).

#### v1.15.0 `2017-07-17`
- `Added`
    - Generate initial AppVeyor configuration for CLI projects. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#196](https://github.com/jonathantorres/construct/issues/196).
- `Fixed`
    - The Xdebug extension disabling has been removed from the Travis CI configuration as Composer takes care of this since 1.3. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#194](https://github.com/jonathantorres/construct/issues/194).
    - Utilise PHPUnit's forward compatibility layer for PHPUnit 6. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#179](https://github.com/jonathantorres/construct/issues/179).
    - The email notification for successful Travis CI builds has been disabled. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#198](https://github.com/jonathantorres/construct/issues/198).
    - The generated change log has been aligned with the [Keep a Changelog](http://keepachangelog.com/) format. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#202](https://github.com/jonathantorres/construct/issues/202).
    - The CLI binary has been moved to `bin` to achieve [PDS](http://php-pds.com) compliance. Done by [@raphaelstolt](https://github.com/raphaelstolt).

#### v1.14.1 `2017-04-07`
- `Added`
    - Make the PHP Coding Standards Fixer linting fail early to speedup builds. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#187](https://github.com/jonathantorres/construct/issues/187).
- `Fixed`
    - The PHP Coding Standards Fixer cache directory is present in the generated Travis CI configuration. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#186](https://github.com/jonathantorres/construct/issues/186).
    - The PHPUnit 6 test skeleton is created on PHP 7.0.* after a fix of the version guard. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#192](https://github.com/jonathantorres/construct/issues/192).
    - The CLI artifacts are constructed as expected. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#190](https://github.com/jonathantorres/construct/issues/190).

#### v1.14.0 `2017-03-29`
- `Added`
    - With the `--cli-framework` option a CLI project can be generated. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#175](https://github.com/jonathantorres/construct/issues/175).
- `Fixed`
    - The package `vlucas/phpdotenv` is added as a non development requirement. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#176](https://github.com/jonathantorres/construct/issues/176).
    - Fix PHP Coding Standards Fixer Composer scripts are set on Windows. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#183](https://github.com/jonathantorres/construct/issues/183).
    - Command integration tests use no output decoration and are green again. Also enabled CI of the command integration tests on Windows via [AppVeyor](https://www.appveyor.com). Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#181](https://github.com/jonathantorres/construct/issues/181).

#### v1.13.1 `2017-03-01`
- `Fixed`
    - Fix generated PHP Coding Standards Fixer configuration. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#172](https://github.com/jonathantorres/construct/issues/172).
    - Executable test skeleton for PHPUnit 6. Done by [@raphaelstolt](https://github.com/raphaelstolt). See [#170](https://github.com/jonathantorres/construct/issues/170).

#### v1.13.0 `2016-12-11`
- `Added`
    - Coding standard checks are only run against a single PHP version. Done by [@raphaelstolt](https://github.com/raphaelstolt).
    - PHP version `7.1` is added to Travis CI builds. Done by [@raphaelstolt](https://github.com/raphaelstolt).

src/Defaults.php
<?php

declare(strict_types = 1);

namespace Construct;

class Defaults
{
    /**
     * The available open source licenses. (more: http://choosealicense.com/licenses)
     *
     * @var array
     */
    private $licenses = ['MIT', 'Apache-2.0', 'GPL-2.0', 'GPL-3.0'];

    /**
     * The available testing frameworks.
     *
     * @var array
     */
    private $testingFrameworks = ['phpunit', 'behat', 'phpspec', 'codeception'];

    /**
     * Available php versions to test on travis.
     *
     * @var array
     */
    private $phpVersions = ['5.6', '7.0', '7.1', '7.2'];

    /**
     * Php versions without a semver scheme.
     *
     * @var array
     */
    private $nonSemverPhpVersions = ['hhvm', 'nightly'];

    /**
     * The default CLI framework.
     *
     * @var string
     */
    private $cliFramework = 'symfony/console';

    /**
     * The default testing framework.
     *
     * @var string
     */
    private $testingFramework = 'phpunit';

    /**
     * The default license.
     *
     * @var string
     */
    private $license = 'MIT';

    /**
     * The default project namespace.
     *
     * @var string
     */
    private $projectNamespace = 'Vendor\Project';

    /**
     * The default configuration file.
     *
     * @var string
     */
    private $configurationFile = '.construct';

    /**
     * Php version currently used on the system.
     *
     * @var string
     */
    private $systemPhpVersion = '';

    /**
     * Initialize Default values.
     *
     * @return void
     */
    public function __construct()
    {
        $this->systemPhpVersion = PHP_MAJOR_VERSION . '.' . PHP_MINOR_VERSION;
    }

    /**
     * Get the available open source licenses.
     *
     * @return array
     */
    public function getLicenses(): array
    {
        return $this->licenses;
    }

    /**
     * Get the supported testing frameworks.
     *
     * @return array
     */
    public function getTestingFrameworks(): array
    {
        return $this->testingFrameworks;
    }

    /**
     * Get the available php versions to test on Travis.
     *
     * @return array
     */
    public function getPhpVersions(): array
    {
        return $this->phpVersions;
    }

    /**
     * Get the php versions without a semver scheme.
     *
     * @return array
     */
    public function getNonSemverPhpVersions(): array
    {
        return $this->nonSemverPhpVersions;
    }

    /**
     * Get the default CLI Framework.
     *
     * @return string
     */
    public function getCliFramework(): string
    {
        return $this->cliFramework;
    }

    /**
     * Get the default testing framework.
     *
     * @return string
     */
    public function getTestingFramework(): string
    {
        return $this->testingFramework;
    }

    /**
     * Get the default license.
     *
     * @return string
     */
    public function getLicense(): string
    {
        return $this->license;
    }

    /**
     * Get the default project namespace.
     *
     * @return string
     */
    public function getProjectNamespace(): string
    {
        return $this->projectNamespace;
    }

    /**
     * Get the default name of the configuration file.
     *
     * @return string
     */
    public function getConfigurationFile(): string
    {
        return $this->configurationFile;
    }

    /**
     * Get the php version currently installed on the system.
     *
     * @return string
     */
    public function getSystemPhpVersion(): string
    {
        return $this->systemPhpVersion;
    }
}
